pkg/webhook/resources/persistentvolumeclaim/mutator.go

```diff
@@ -47,32 +48,86 @@ func (m *pvcMutator) Create(_ *types.Request, newObj runtime.Object) (types.Patc
 	pvc := newObj.(*corev1.PersistentVolumeClaim)
 
 	logrus.Debugf("create PVC %s/%s with mutator", pvc.Namespace, pvc.Name)
+	patchOPs := []string{}
+	isGoldenImage := false
 
 	// check pvc is related to the vm image
-	_, err := m.vmImageCache.Get(pvc.Namespace, pvc.Name)
+	patchOp, err := m.patchGoldenImageAnnotation(pvc)
 	if err != nil {
+		return nil, err
+	}
+	if patchOp != "" {
+		isGoldenImage = true
+		patchOPs = append(patchOPs, patchOp)
+	}
+
+	// patch populator
+	if !isGoldenImage {
+		patchOp, err = m.patchDataSource(pvc)
+		if err != nil {
+			return nil, err
+		}
+		patchOPs = append(patchOPs, patchOp)
+	}
+
+	// no need to patch
+	if len(patchOPs) == 0 {
+		return nil, nil
+	}
+	return patchOPs, nil
+}
+
+func (m *pvcMutator) patchGoldenImageAnnotation(pvc *corev1.PersistentVolumeClaim) (string, error) {
+	if _, err := m.vmImageCache.Get(pvc.Namespace, pvc.Name); err != nil {
 		if apierrors.IsNotFound(err) {
 			logrus.Infof("PVC %s/%s is not related to the VM image, skip patch", pvc.Namespace, pvc.Name)
-			return nil, nil
+			return "", nil
 		}
-		return nil, err
+		return "", err
 	}
 
 	annotations := pvc.GetAnnotations()
 
-	if _, find := annotations[util.AnnotationGoldenImage]; find {
-		if annotations[util.AnnotationGoldenImage] == "true" {
-			return nil, nil
-		}
+	if v, find := annotations[util.AnnotationGoldenImage]; find && v == "true" {
+		return "", nil
 	}
 	annotations[util.AnnotationGoldenImage] = "true"
 
 	annoVal, err := json.Marshal(annotations)
 	if err != nil {
 		logrus.Warnf("failed to marshal annotations: %v, err: %v", annotations, err)
-		return nil, err
+		return "", err
+	}
+
+	// patch annotation
+	return fmt.Sprintf(`{"op": "replace", "path": "/metadata/annotations", "value": %s}`, string(annoVal)), nil
+}
+
+func (m pvcMutator) patchDataSource(pvc *corev1.PersistentVolumeClaim) (string, error) {
+	// do no-op if volume mode is block
+	if pvc.Spec.VolumeMode != nil && *pvc.Spec.VolumeMode == corev1.PersistentVolumeBlock {
+		logrus.Debugf("PVC %s/%s is block volume, skip patch", pvc.Namespace, pvc.Name)
+		return "", nil
+	}
+	annotations := pvc.GetAnnotations()
+	if v, find := annotations[util.AnnotationVolForVM]; !find || v != "true" {
+		logrus.Debugf("PVC %s/%s is not for VM, skip patch", pvc.Namespace, pvc.Name)
+		return "", nil
+	}
+
+	cdiAPIGroup := cdicommon.AnnAPIGroup
+	dataSourceRef := corev1.TypedLocalObjectReference{
+		APIGroup: &cdiAPIGroup,
+		Kind:     util.KindVolumeImportSource,
+		Name:     util.ImportSrourceBlank,
```

Comment:
```suggestion
		Name:     util.ImportSourceBlank,
```

Reply:
I updated this Name with the on-demand creation.